### PR TITLE
Add call caching information to metadata endpoint.  Closes #559

### LINKED ISSUE #559: Add call caching fields to metadata endpoing
At least "allows result reuse" and "results cloned" (the latter being transformed from an fk to a boolean). Returning hashes might be helpful too for diagnosing why executions don't avoid.

### PR DESCRIPTION
**Pull request by scottfrazer**
see `calls['w.x']['cache']` and `calls['w.y']['cache']`:

```json
HTTP/1.1 200 OK
Content-Length: 2056
Content-Type: application/json; charset=UTF-8
Date: Wed, 23 Mar 2016 13:22:35 GMT
Server: spray-can/1.3.2

{
    "calls": {
        "w.x": [
            {
                "attempt": 1,
                "backend": "Local",
                "cache": {
                    "allowResultReuse": true
                },
                "end": "2016-03-23T09:22:30.000-04:00",
                "executionEvents": [],
                "executionStatus": "Done",
                "inputs": {
                    "s": "foo"
                },
                "outputs": {
                    "t": "foo"
                },
                "returnCode": 0,
                "runtimeAttributes": {
                    "continueOnReturnCode": "0",
                    "docker": "ubuntu:latest",
                    "failOnStderr": "false"
                },
                "shardIndex": -1,
                "start": "2016-03-23T09:22:28.000-04:00",
                "stderr": "/Users/sfrazer/projects/cromwell/cromwell-executions/w/d21f9706-0d85-4859-af96-cb58b6913085/call-x/stderr",
                "stdout": "/Users/sfrazer/projects/cromwell/cromwell-executions/w/d21f9706-0d85-4859-af96-cb58b6913085/call-x/stdout"
            }
        ],
        "w.y": [
            {
                "attempt": 1,
                "backend": "Local",
                "cache": {
                    "allowResultReuse": true,
                    "cacheHitCall": "w.x",
                    "cacheHitWorkflow": "d21f9706-0d85-4859-af96-cb58b6913085"
                },
                "end": "2016-03-23T09:22:30.000-04:00",
                "executionEvents": [],
                "executionStatus": "Done",
                "inputs": {
                    "s": "foo"
                },
                "outputs": {
                    "t": "foo"
                },
                "returnCode": 0,
                "runtimeAttributes": {
                    "continueOnReturnCode": "0",
                    "docker": "ubuntu:latest",
                    "failOnStderr": "false"
                },
                "shardIndex": -1,
                "start": "2016-03-23T09:22:30.000-04:00",
                "stderr": "/Users/sfrazer/projects/cromwell/cromwell-executions/w/d21f9706-0d85-4859-af96-cb58b6913085/call-y/stderr",
                "stdout": "/Users/sfrazer/projects/cromwell/cromwell-executions/w/d21f9706-0d85-4859-af96-cb58b6913085/call-y/stdout"
            }
        ]
    },
    "end": "2016-03-23T09:22:31.000-04:00",
    "id": "d21f9706-0d85-4859-af96-cb58b6913085",
    "inputs": {},
    "outputs": {
        "w.x.t": "foo",
        "w.y.t": "foo"
    },
    "start": "2016-03-23T09:22:28.000-04:00",
    "status": "Succeeded",
    "submission": "2016-03-23T09:22:28.000-04:00",
    "workflowName": "w"
}
```

### CODE REVIEW

**Comment by geoffjentry** [comment]
I'm having trouble seeing it on my phone but how many fields is that metadata case class up to now?

Not here, as it'd be a breaking change but we'll need to discuss subsections or something soon

File: engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala [hunk 1]
```diff
@@ -171,7 +171,9 @@ class WorkflowDescriptorSpec extends FlatSpec with Matchers with ScalaFutures {
           1,
           Map.empty,
           None,
-          None)
+          None,
+          false,
+          false)
         call -> Seq(metadata)
     }).toMap
```
Comment (mcovarr) [inline engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala]: There doesn't seem to be a substantive test that these fields are correctly filled out for hits / non-hits and ALLOW_RESULTS_REUSE turned on or not.

Comment (scottfrazer) [inline engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala]: Yeah, there isn't.  I'm actually struggling to find where our metadata endpoint tests are in general...  Do you know where they are?

**Comment by mcovarr** [comment]
@geoffjentry actually the constructor is at arity 20 now.  I think the point Jeff is driving at is that there aren't built-in Spray marshallers for arities > 22, and I'm actually not sure this won't break down completely for arities > 22 (it is possible to have case class constructors of arity > 22, but functions can't have arity > 22 so something else might still break).

So beyond the aesthetic issues here this has become like a game of musical chairs; it might be nice to address this proactively.

File: engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala (hunk 1 above)
Comment (mcovarr) [inline engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala]: hopefully not only in my imagination... let me go look

**Comment by geoffjentry** [comment]
I put that tilde there in case some pedant came and demonstrated I was off by a bit. Looks like it was a good idea.

Yeah, that's exactly what I mean.

However as this has impact on our clients (one of whom is currently deeply concerned about breaking changes at the moment) I think this needs to involve discussion with other folks prior to making an executive decision

File: engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala (hunk 1 above)
Comment (mcovarr) [inline engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala]: Actually there are some nice tests in `RetryableCallsSpec` and `WorkflowManagerActorSpec`.

Comment (scottfrazer) [inline engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala]: I think there are a few meta-data related tests in WorkflowManagerActorSpec, I'll see if I can add it there

File: engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala [hunk 2]
```diff
@@ -126,6 +127,7 @@ class WorkflowDescriptorSpec extends FlatSpec with Matchers with ScalaFutures {
       Map.empty,
       None)
 
+    implicit val patienceConfig = PatienceConfig(timeout = Span(10, Seconds), interval = Span(15, Millis))
```
Comment (scottfrazer) [inline engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala]: @Horneth perhaps try adding this to your branch if you push any more changes to see if this helps with your test failure too?

Comment (Horneth) [inline engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala]: @scottfrazer I will thx !

File: engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala (hunk 1 above)
Comment (scottfrazer) [inline engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala]: Test added

**Comment by Horneth** [comment]
Would it be possible to have the Workflow ID / Call FQN of the call that was used as a cache in the metadata ?

**Comment by scottfrazer** [comment]
@Horneth that is a good idea.

**Comment by cjllanwarne** [comment]
@Horneth - Yeah! That's actually a great idea

**Comment by cjllanwarne** [comment]
@geoffjentry @mcovarr yeah we will need to redesign call metadata. Actually IIRC (full disclosure, I might NOT remember correctly) a few fields in there now are backend-specific so maybe they'll disappear away soon?

**Comment by geoffjentry** [comment]
@cjllanwarne We need to debulk it by more than just a few :) One of the cited reasons for them stopping the max arity where they did was that if you're anywhere close you're doing something wrong.

**Comment by geoffjentry** [comment]
@Horneth another plus-one on your good idea

**Comment by mcovarr** [comment]
How about rolling your two new fields up with @Horneth's new field idea and making a composite object that has all the caching stuff in one logical grouping that only increases arity by 1 instead of 3?  :smile:

**Comment by scottfrazer** [comment]
@mcovarr sounds good

**Comment by mcovarr** [comment]
Do you think there would be value in showing whether a call is used as the cached result for other calls?

**Comment by scottfrazer** [comment]
@mcovarr I guess that could be useful, though perhaps that's something we should wait for people to ask for?  Or perhaps a separate ticket?  Seems like it be just an additive feature

**Comment by scottfrazer** [comment]
Wheel chooses @kshakir (seems appropriate) and @mcovarr

File: engine/src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala [hunk 3]
```diff
@@ -890,4 +890,16 @@ class SlickDataAccess(databaseConfig: Config) extends DataAccess {
...
+
+  override def callCacheDataByExecution(id: WorkflowId)(implicit ec: ExecutionContext): Future[Traversable[ExecutionWithCacheData]] = {
+    val action = dataAccess.executionsWithCacheHitWorkflowAndCall(id.toString).result
+    runTransaction(action).map(_.map {
+      case (execution, Some(cacheHitWorkflowUuid), Some(cacheHitCallFqn)) =>
+        ExecutionWithCacheData(execution, Option(CallCacheHit(cacheHitWorkflowUuid, cacheHitCallFqn)))
+      case (execution, None, None) =>
+        ExecutionWithCacheData(execution, None)
+      case (execution, cacheHitWorkflowUuid, cacheHitCallFqn) =>
+        throw new IllegalStateException(s"Expecting no cache hit or (workflow UUID, call FQN) tuple.  Got ($cacheHitWorkflowUuid, $cacheHitCallFqn)")
```
Comment (scottfrazer) [inline engine/src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala]: Anybody know a better way to handle this case?  Will this turn into a `Future.failure`?

Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala]: yes this should fail the Future

File: engine/src/main/scala/cromwell/engine/db/slick/ExecutionComponent.scala [hunk 4]
```diff
@@ -161,6 +164,17 @@ trait ExecutionComponent {
       if execution.executionId === executionId
     } yield (execution.status, execution.rc))
 
+  val executionsWithCacheHitWorkflowAndCall = Compiled(
+    (workflowExecutionUuid: Rep[String]) => for {
+      (((e, e1), we1), we) <- executions joinLeft executions on (_.resultsClonedFrom === _.executionId) joinLeft workflowExecutions on {
+        case ((e, maybeE1), we1) => maybeE1.map(_.workflowExecutionId) === we1.workflowExecutionId
+      } joinLeft workflowExecutions on {
+        case (((e, maybeE1), maybeWe1), we) => e.workflowExecutionId === we.workflowExecutionId
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/db/slick/ExecutionComponent.scala]: looks like you can underscore all the maybes

Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/db/slick/ExecutionComponent.scala]: a few comments here for posterity perhaps?  :smile:

Comment (scottfrazer) [inline engine/src/main/scala/cromwell/engine/db/slick/ExecutionComponent.scala]: StackOverflow links?  Corresponding SQL query?  both?

Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/db/slick/ExecutionComponent.scala]: The `we` on 172 is not the same `we` of line 169 and 174 fame, which is a little confusing at first

File: engine/src/main/scala/cromwell/engine/db/slick/ExecutionInfoComponent.scala [hunk 5]
```diff
@@ -5,7 +5,6 @@ case class ExecutionInfo(executionId: Int,
                          value: Option[String] = None,
                          executionInfoId: Option[Int] = None)
 
-case class ExecutionAndExecutionInfo(execution: Execution, executionInfo: ExecutionInfo)
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/db/slick/ExecutionInfoComponent.scala]: yay

File: engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala [hunk 6]
```diff
@@ -14,6 +14,7 @@ import org.scalatest.concurrent.PatienceConfiguration.Timeout
 import org.scalatest.concurrent.ScalaFutures
 import org.scalatest.prop.TableDrivenPropertyChecks._
 import org.scalatest.prop.Tables.Table
+import org.scalatest.time.{Millis, Seconds, Span}
 import org.scalatest.time.SpanSugar._
 import org.scalatest.time.{Millis, Seconds, Span}
```
Comment (mcovarr) [inline engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala]: imports so nice you've imported them twice

File: engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala [hunk 7]
```diff
@@ -173,8 +174,10 @@ class WorkflowDescriptorSpec extends FlatSpec with Matchers with ScalaFutures {
           Seq.empty,
           1,
           Map.empty,
-          None,
-          None)
+          preemptible=None,
+          cache=None,
+          failures=None
```
Comment (mcovarr) [inline engine/src/test/scala/cromwell/engine/WorkflowDescriptorSpec.scala]: yay keywords, spacing could use a bit of cleanup

File: engine/src/main/scala/cromwell/webservice/WorkflowJsonSupport.scala [hunk 8]
```diff
@@ -37,7 +38,8 @@ object WorkflowJsonSupport extends DefaultJsonProtocol {
   implicit val unqualifiedFailureEventEntry = jsonFormat2(FailureEventEntry)
   implicit val qualifiedFailureEventEntry = jsonFormat4(QualifiedFailureEventEntry)
   implicit val executionEventProtocol = jsonFormat3(ExecutionEventEntry)
-  implicit val callMetadataProtocol = jsonFormat18(CallMetadata)
+  implicit val callCacheHitProtocol = jsonFormat3(CallCacheData)
+  implicit val callMetadataProtocol = jsonFormat19(CallMetadata)
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/webservice/WorkflowJsonSupport.scala]: yay back down to a svelte 19 fields

**Comment by mcovarr** [comment]
:+1: just a few nits, don't worry about that feature suggestion

File: engine/src/main/scala/cromwell/engine/db/slick/ExecutionComponent.scala (hunk 4 above)
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/db/slick/ExecutionComponent.scala]: I was thinking just a verbal description of what's going on here :smile:

**Comment by kshakir** [comment]
:+1:

**Pull request merged**